File: tools/generators/src/mirror_list_generator.js
import yaml from 'js-yaml';
import fs from 'fs';
import {COMMON_PLATFORM, generateLink, HOST} from './url_util.js';

const DATA_VERSIONS_PATH = 'data/versions.yml';
const MIRROR_LIST_CONFIGS_PATH = 'data/mirrorlist_configs.yml';
const EXPORT_PATH = 'content/en/mirrorlist';

(function main() {
  const versions = loadVersions();

  for (let i = 0; i < versions.length; i++) {
    const version = versions[i];

    writeMirrorList(version, {mirrors: createMirrorList(version, false)}, false);
    writeMirrorList(version, {mirrors: createMirrorList(version, true)}, true);

    if (version.releases) {
      for (let j = 0; j < version.releases.length; j++) {
        const release = version.releases[j];
        const releaseName = release.release_version ?? version.name;
        const innerVersion = {
          name: releaseName,
          flavor: release.name,
        }

        writeMirrorList(innerVersion, {mirrors: createMirrorList(innerVersion, false)}, false);
        writeMirrorList(innerVersion, {mirrors: createMirrorList(innerVersion, true)}, true);
      }
    }
  }
})();

/**
 * @param {Version} version
 * @param {Object} mirrorList
 * @param {boolean} isMono
 */
function writeMirrorList(version, mirrorList, isMono) {
  let fileName = `${version.name}.${version.flavor}`;

  if (version.name === '3.0') {
    fileName = `${version.name}-${version.flavor}`;
  }

  if (isMono) {
    fileName = `${fileName}.mono`;
  }

  fs.writeFile(`${EXPORT_PATH}/${fileName}.json`, JSON.stringify(mirrorList), (err) => {
    if (err) throw err;
  });
}


/**
 * @param {Object} version
 * @param {string} version.name
 * @param {string} version.flavor
 * @param {boolean} isMono
 */
function createMirrorList(version, isMono) {
  const mirrors = []
  const mirrorListConfigs = loadMirrorListConfigs();
  const mirrorListHosts = mirrorListConfigs.hosts;

  const versionBits = version.name.split('.');
  const versionMajMin = `${versionBits[0]}.${versionBits[1]}`;
  const versionDefaults = mirrorListConfigs.defaults.find((mirror) => mirror.name === versionMajMin);

  if (versionDefaults) {
    const mirrorHosts = version.flavor !== "stable" ? versionDefaults.preview : versionDefaults.stable;

    for (let i = 0; i < mirrorHosts.length; i++) {
      const hostName = mirrorHosts[i];
      const mirrorHost = mirrorListHosts.find((host) => host.name === hostName);
      const mirrorUrl = generateLink(version, COMMON_PLATFORM.templates, isMono, hostName);

      if (mirrorUrl !== '#') {
        const mirror = {
          name: mirrorHost.title,
          url: mirrorUrl,
        }
        mirrors.push(mirror);
      }
    }
  } else {
    const mirrorHost = mirrorListHosts.find((host) => host.name === HOST.tuxfamily);
    const mirrorUrl = generateLink(version, COMMON_PLATFORM.templates, isMono, HOST.tuxfamily);

    if (mirrorUrl !== '#') {
      const mirror = {
        name: mirrorHost.title,
        url: mirrorUrl,
      }
      mirrors.push(mirror);
    }
  }

  return mirrors;
}


/**
 * @typedef {Object} MirrorHost
 * @property {string} name
 * @property {string} title
 */

/**
 * @typedef {Object} Mirror
 * @property {string} name
 * @property {string[]} stable
 * @property {string[]} preview
 */

/**
 * @typedef {Object} MirrorListConfig
 * @property {MirrorHost[]} hosts
 * @property {Mirror[]} defaults
 */

/**
 * @returns {MirrorListConfig}
 */
function loadMirrorListConfigs() {
  const fileContents = fs.readFileSync(MIRROR_LIST_CONFIGS_PATH, 'utf8');
  return yaml.load(fileContents);
}

/**
 * @typedef {Object} VersionRelease
 * @property {string} name
 * @property {string} release_date
 * @property {string} release_notes
 */

/**
 * @typedef {Object} Version
 * @property {string} name
 * @property {string} falvor
 * @property {string} release_date
 * @property {string} release_notes
 * @property {boolean} featured
 * @property {VersionRelease[]} releases
 */

/**
 * @returns {Version[]}
 */
function loadVersions() {
  const versions = fs.readFileSync(DATA_VERSIONS_PATH, 'utf8')
  return yaml.load(versions);
}

File: _plugins/mirrorlist_generator.rb
module MirrorlistGeneratorPlugin
	class MirrorlistGenerator < Jekyll::Generator
	  safe true

	  def generate(site)
		puts "Generating mirrorlist pages..."

		site.data["versions"].each do |version|
		  version_name = version["name"]
		  puts "    Rendering version '#{version_name}'."

		  # Generate files for the current main flavor (stable or latest pre-release).
		  site.pages << MirrorlistPage.new(site, version, false)
		  site.pages << MirrorlistPage.new(site, version, true)

		  # Iterate through previous releases and generate files for them as well.
		  if version.key?("releases")
			version["releases"].each do |release|
				version_name = release.key?("release_version") ? release["release_version"] : version["name"]

				prerelease = {
					'name' => version_name,
					'flavor' => release["name"]
				}
				site.pages << MirrorlistPage.new(site, prerelease, false)
				site.pages << MirrorlistPage.new(site, prerelease, true)
			end
		  end
		end

		puts "Finished generating the mirrorlist."
	  end
	end

	class MirrorlistPage < Jekyll::Page
	  def initialize(site, version, mono)
		# Configure static information and path.
		@site = site
		@base = site.source
		@dir  = 'mirrorlist'

		# Generate the version identificator.
		version_name = version["name"]
		version_flavor = version["flavor"]

		version_id = "#{version_name}.#{version_flavor}"
		if version_name == "3.0" # Hack for the only version before the naming scheme has been changed.
			version_id = "#{version_name}-#{version_flavor}"
		end
		if mono
			version_id = "#{version_id}.mono"
		end

		version_bits = version_name.split(".")
		version_majmin = "#{version_bits[0]}.#{version_bits[1]}"

		# Configure the permalink information.
		@basename = version_id
		@ext      = '.json'
		@name     = "#{version_id}.json"

		# Generate the list of mirrors.
		mirrors = []

		mirrorlist_configs = site.data["mirrorlist_configs"]
		mirrorlist_hosts = mirrorlist_configs["hosts"]

		version_defaults = mirrorlist_configs["defaults"].find { |config| config["name"] == version_majmin }

		unless version_defaults.nil?
			mirror_hosts = version_defaults["stable"]
			unless version_flavor == "stable"
				mirror_hosts = version_defaults["preview"]
			end

			mirror_hosts.each do |host_name|
				mirror_host = mirrorlist_hosts.find { |host| host["name"] == host_name }
				mirror_url = make_download(version, mono, host_name)

				unless mirror_url == "#"
					mirror = { 'name' => mirror_host["title"], 'url' => mirror_url }
					mirrors.push(mirror)
				end
			end
		else
			mirror_host = mirrorlist_hosts.find { |host| host["name"] == "tuxfamily" }
			mirror_url = make_download(version, mono, "tuxfamily")

			unless mirror_url == "#"
				mirror = { 'name' => mirror_host["title"], 'url' => mirror_url }
				mirrors.push(mirror)
			end
		end

		# Initialize data hash to pass objects to the template.
		@data = {
		  'version_mirrors' => mirrors
		}

		# Defaults can be configured via the `_config.yml`, same as with normal
		# collections.
		data.default_proc = proc do |_, key|
		  site.frontmatter_defaults.find(relative_path, :mirrorlist, key)
		end
	  end

	  # This is a hack to generate the URL from code. I don't know a better way to hook into Liquid and use its
	  # registered filters. But it's short and sweet enough for now (or for... ever).
	  def make_download(version, mono, host)
		template = Liquid::Template.parse('{{ version | make_download: "templates", mono, host }}')

		assigns = {
			'version' => version,
			'mono' => mono,
			'host' => host
		}
		# We need to pass the site reference to registers because make_downloads relies on this context being
		# present.
		return template.render(assigns, registers: { site: @site })
	  end

	  def url_placeholders
		{
		  :path       => @dir,
		  :basename   => basename,
		  :output_ext => output_ext,
		}
	  end
	end
  end
